src/platform/ModulesManager.c

```diff
@@ -0,0 +1,742 @@
+static const char* g_componentName = "ComponentName";
+static const char* g_objectName = "ObjectName";
+
+typedef struct MODULE_SESSION
+{
+    MODULE* module;
+    MMI_HANDLE handle;
+
+    struct MODULE_SESSION* next;
+} MODULE_SESSION;
+
+typedef struct SESSION
+{
+    char* uuid;
+    char* client;
+    MODULE_SESSION* modules;
+
+    struct SESSION* next;
+} SESSION;
+
+typedef struct REPORTED_OBJECT
+{
+    char* component;
+    char* object;
+} REPORTED_OBJECT;
+
+static SESSION* g_sessions = NULL;
+static MODULE* g_modules = NULL;
+static REPORTED_OBJECT* g_reported = NULL;
+static int g_reportedTotal = 0;
+
+OSCONFIG_LOG_HANDLE g_platformLog = NULL;
+
+OSCONFIG_LOG_HANDLE GetPlatformLog(void)
+{
+    return g_platformLog;
+}
+
+static void LoadModules(const char* directory, const char* configJson)
+{
+    MODULE* module = NULL;
+    DIR* dir = NULL;
+    struct dirent* entry = NULL;
+    char* path = NULL;
+    char* clientName = NULL;
+    int version = 0;
+    JSON_Value* config = NULL;
+    JSON_Object* configObject = NULL;
+    JSON_Array* reportedArray = NULL;
+    JSON_Object* reportedObject = NULL;
+    REPORTED_OBJECT* reported = NULL;
+    int reportedCount = 0;
+    int reportedTotal = 0;
+    int i = 0;
+    ssize_t clientNameSize = 0;
+    ssize_t reportedSize = 0;
+    ssize_t pathSize = 0;
+
+    OsConfigLogInfo(GetPlatformLog(), "Loading modules...");
+
+    if ((NULL == directory) || (NULL == configJson))
+    {
+        OsConfigLogError(GetPlatformLog(), "LoadModules(%p, %p) called with invalid arguments", directory, configJson);
+    }
+    else if(NULL == (dir = opendir(directory)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to open module directory: %s", directory);
+    }
+    else if (NULL == (config = json_parse_file(configJson)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to parse configuration JSON (%s)", configJson);
+    }
+    else if (NULL == (configObject = json_value_get_object(config)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to get config object");
+    }
+    else if (0 == (version = json_object_get_number(configObject, g_modelVersion)))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to get model version from configuration JSON (%s)", configJson);
+    }
+    else
+    {
+        // "Azure OSConfig <version>;<osconfig version>" + null-terminator
+        clientNameSize = strlen(AZURE_OSCONFIG) + strlen(OSCONFIG_VERSION) + 5;
+
+        if (NULL == (clientName = (char*)malloc(clientNameSize)))
+        {
+            OsConfigLogError(GetPlatformLog(), "Failed to allocate memory for client name");
+        }
+        else
+        {
+            memset(clientName, 0, clientNameSize);
+            snprintf(clientName, clientNameSize, "%s %d;%s", AZURE_OSCONFIG, version, OSCONFIG_VERSION);
+        }
+
+        OsConfigLogInfo(GetPlatformLog(), "Loading modules from '%s' (%s)", directory, clientName);
```

Comment:
>OsConfigLogInfo(GetPlatformLog(), "Loading modules from '%s' (%s)", directory, clientName); [](http://example.com/codeflow?start=8&length=91)

This repeats the text from line 72. Combine the 2 traces in 1 or change this one's text